xds/src/main/java/io/grpc/xds/sds/SslContextSecretVolumeSecretProvider.java

```diff
@@ -0,0 +1,164 @@
+    implements SecretProvider<SslContext> {
+  @VisibleForTesting
+  static final Logger logger = Logger
+      .getLogger(SslContextSecretVolumeSecretProvider.class.getName());
+
+  @VisibleForTesting
+  boolean server;
+  @VisibleForTesting
+  String privateKey;
+  @VisibleForTesting
+  String privateKeyPassword;
+  @VisibleForTesting
+  String certificateChain;
+  @VisibleForTesting
+  String trustedCa;
+
+  SslContextSecretVolumeSecretProvider(TlsCertificate tlsCertificate,
+      CertificateValidationContext certContext, boolean server) {
+    // first validate
+    if (server) {
+      validateTlsCertificate(tlsCertificate, false);
+      // certContext exists in case of mTLS, else null for a server
+      if (certContext != null) {
+        certContext = validateCertificateContext(certContext, true);
+      }
+    } else {
+      validateCertificateContext(certContext, false);
+      // tlsCertificate exists in case of mTLS, else null for a client
+      if (tlsCertificate != null) {
+        tlsCertificate = validateTlsCertificate(tlsCertificate, true);
+      }
+    }
+    if (tlsCertificate != null) {
+      this.privateKey = tlsCertificate.getPrivateKey().getFilename();
+      if (tlsCertificate.hasPassword()) {
+        this.privateKeyPassword = tlsCertificate.getPassword().getInlineString();
+      }
+      this.certificateChain = tlsCertificate.getCertificateChain().getFilename();
+    }
+    if (certContext != null) {
+      this.trustedCa = certContext.getTrustedCa().getFilename();
+    }
+    this.server = server;
+  }
+
+  @VisibleForTesting
+  static CertificateValidationContext validateCertificateContext(
+      CertificateValidationContext certContext, boolean optional) {
+    if (certContext == null || !certContext.hasTrustedCa()) {
+      checkArgument(optional, "certContext is required");
+      return null;
+    }
+    checkArgument(certContext.getTrustedCa().getSpecifierCase()
+            == SpecifierCase.FILENAME,
+        "filename expected");
+    return certContext;
+  }
+
+  @VisibleForTesting
+  static TlsCertificate validateTlsCertificate(TlsCertificate tlsCertificate,
+      boolean optional) {
+    if (tlsCertificate == null) {
+      checkArgument(optional, "tlsCertificate is required");
+      return null;
+    }
+    if (optional && (tlsCertificate.getPrivateKey().getSpecifierCase()
+        == SpecifierCase.SPECIFIER_NOT_SET)
+            && (tlsCertificate.getCertificateChain().getSpecifierCase()
+            == SpecifierCase.SPECIFIER_NOT_SET)) {
+      return null;
+    }
+    checkArgument(
+        tlsCertificate.getPrivateKey().getSpecifierCase() == SpecifierCase.FILENAME,
+        "filename expected");
+    checkArgument(
+        tlsCertificate.getCertificateChain().getSpecifierCase() == SpecifierCase.FILENAME,
+        "filename expected");
+    return tlsCertificate;
+  }
+
+  @Override
+  public void addCallback(final Callback<SslContext> callback, Executor executor) {
```

Comment:
will this consume by ProtocolNegotiator?

Reply:
correct

Reply:
discussed offline, we want to add a comment because this evolves IO which can block `EventLoop`.

Reply:
done